content/blog/2018-03-22-monkeydo.md

```diff
@@ -0,0 +1,648 @@
+
+If a dataframe is supplied, there is now a `.keep_all` option which allows for all columns of the input to be retained, not just the column that contains the text to be classified. This makes the `monkey_` functions work even more like a `mutate()`; rather than returning an object that has to be joined on the original input, we do that association for the user.
+
+
+
+```r
+sw <- dplyr::starwars %>% 
+  dplyr::select(name, height) %>% 
+  dplyr::sample_n(length(text_w_empties))
+
+df <- tibble::tibble(text = text_w_empties) %>% 
+  dplyr::bind_cols(sw)
+
+df %>% monkey_classify(text, unnest = FALSE, .keep_all = TRUE, verbose = FALSE)
+```
+
+```
+## # A tibble: 5 x 4
+##   name             height text                                     res    
+##   <chr>             <int> <chr>                                    <list> 
+## 1 Owen Lars           178 It is a truth universally acknowledged,… <data.…
+## 2 Jek Tono Porkins    180 When Mr. Bilbo Baggins of Bag End annou… <data.…
+## 3 Bib Fortuna         180 ""                                       <data.…
+## 4 Dormé               165 I'm not an ambiturner. I can't turn lef… <data.…
+## 5 Jocasta Nu          167 " "                                      <data.…
+```
+
+We see that the input column, `text` is sandwiched between the other columns of the original dataframe (the Starwars ones) and the output column `res`.
+
+The hope is that all of this serves to improve the data safety and user experience of the package.
+
+
+#### Developing functions in tandem
+
+Something I've been thinking about while working on the twin functions `monkey_extract()` and `monkey_classify()` is what the best practice is for developing very similar functions in sync with one another. These two functions are different enough to have different default values (for example, `monkey_extract()` has a default `extractor_id` while `monkey_classify()` has a default `classifier_id`) but are so similar in other regards as to be sort of embarrassingly parallel.
+
+What I've been turning over in my head is the question of how in sync these functions should be during development. As soon as you make a change to one function, should you immediately make the same change to the other? Or is it instead better to work on one function at a time, and, at some checkpoints then batch these changes over to the other function in a big copy-paste job? I've been tending toward the latter but it's seemed a little dangerous to me.
+
+Since there are only two functions to worry about here, creating a function factory to handle them seemed like overkill, but might technically be the best practice. I'd love to hear people's thoughts on how they go about navigating this facet of package development.
```

Comment:
Really like the tone here, and inviting people to comment. Remind me when it's time to publish that we should select some key themes or quotes for tweets to help people see what's in here if they take the time to read it

Reply:
Oh yeah! Tweeting out the question of how best to develop twin functions in tandem might be a good way to get the conversation going